cilium/CFP-39876-clustermesh-filtered-export.md

```diff
@@ -0,0 +1,211 @@
+
+### ClusterMesh API Server Changes
+
+#### Process annotations 
+We will add new watch on namespaces to check for global or local annotations and maintain internal sets for the same. Once we have the list, we will trigger a full sync operation to update the export based on the provided annotation. 
+
+#### Export Filtering
+Only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 
+
+#### ServiceExport controller 
+Provide warning message when a ServiceExport is triggered for a service residing in local namespace 
+
+#### Existing Cilium Resource Types
+- CiliumEndpoint(Slice)
+- CiliumIdentities
+- Global Services
+- MCS ServiceExport
+
+## Impacts / Key Questions
+---
+
+### Impact: Breakage during upgrade
+- The changes are expected to be backward compatible in the sense that the new scoped export logic only takes effect once the user has explicitly annotated namespaces as global or local. As a result, existing clusters will continue to function as before until the user opts into the new behavior. The `clustermesh-default-global-namespace` config option will determine the default behavior for unannotated namespaces.
+---
+
+### Impact: Debugging multi-cluster environment connectivity may become more difficult
+- Failures due to missing identity or label exports may be non-obvious and require deeper inspection into the global/local namespace configuration.
+- Cross-cluster connectivity issues will now depend not only on service presence or routes but also on namespace-level annotations and default behaviors requiring additional verifications on the value of config `clustermesh-default-global-namespace` along with the actual annotations present on the namespaces during debugging.
+- If an application is sending traffic to a destination in another cluster and the traffic is dropped due to policy (as seen in hubble observe), users should inspect whether the source and destination namespaces are marked as global.
+---
+### Alternative Approach 1
+
+### Exporting CiliumEndpoint and CiliumIdentity for Global Services
+
+By default, all endpoints fronted by **global services** are expected to have their `CiliumEndpoint` and `CiliumIdentity` information shared across clusters to support cross-cluster load balancing. If we can determine whether a given `CiliumEndpoint` or `CiliumIdentity` is fronted by a global service, we can use that as the basis for deciding whether to export it.
+
+To enable clients of global services to communicate with their backend service endpoints, we also need a mechanism for users to explicitly mark certain pods as **clients** of global services. This can be achieved through a pod-level annotation.
+
+Since the `clustermesh-apiserver` already watches `CiliumEndpoints` and `CiliumIdentities`, we can maintain in-memory caches to support export decisions:
+
+- **`IPToService`**: a mapping from IP addresses to services that front them.
```

Comment:
is this backend ip to cluster/lb ip?

Reply:
this is the mapping of 

CiliumEndpoint IP ----> List of Services fronting that front that CiliumEndpoint